core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,195 @@
+        if (closed) {
+            return;
+        }
+        closed = true;
+        source.abort();
+        if (inputStream != null) {
+            inputStream.close();
+        }
+    }
+
+    @Override
+    public int available() throws IOException {
+        ensureOpen();
+        if (inputStream == null) {
+            return 0;
+        }
+        return inputStream.available();
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed) {
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private void ensureSubscribed() {
+        if (subscribed.compareAndSet(false, true)) {
+            source.subscribe(subscriber);
+        }
+        subscriber.whenSubscribed.join();
+    }
+
+    private static final class StreamMessageInputStreamSubscriber<T> implements Subscriber<T> {
+
+        private static final InputStream EMPTY_STREAM = new InputStream() {
+            @Override
+            public int read() {
+                return -1;
+            }
+        };
+
+        private final Function<? super T, ? extends HttpData> httpDataConverter;
+        @Nullable
+        private Subscription upstream;
+        private final CompletableFuture<Void> whenSubscribed = new CompletableFuture<>();
+        private final BlockingQueue<InputStream> queue = new LinkedBlockingDeque<>();
+        private volatile boolean closed;
+
+        StreamMessageInputStreamSubscriber(Function<? super T, ? extends HttpData> httpDataConverter) {
+            requireNonNull(httpDataConverter, "httpDataConverter");
+            this.httpDataConverter = httpDataConverter;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            upstream = subscription;
+            whenSubscribed.complete(null);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            if (closed) {
+                 StreamMessageUtil.closeOrAbort(item);
+                 return;
+            }
+            try {
+                queue.add(httpDataConverter.apply(item).toInputStream());
+            } catch (Throwable ex) {
+                StreamMessageUtil.closeOrAbort(item, ex);
+                upstream.cancel();
+                onError(ex);
+            }
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            closed = true;
+            queue.add(EMPTY_STREAM); // to wake the BlockingQueue
+        }
+
+        @Override
+        public void onComplete() {
+            closed = true;
+            queue.add(EMPTY_STREAM); // to wake the BlockingQueue
+        }
+
+        public InputStream nextStream() throws InterruptedException {
+            if (closed) {
+                return EMPTY_STREAM;
+            }
+            upstream.request(1);
+            return queue.take();
+        }
```

Comment:
```java
// AS-IS
public InputStream nextStream() throws InterruptedException {
    if (closed) {
        return EMPTY_STREAM;
    }
    if (queue.isEmpty()) { // request(1) only when queue is empty..
        upstream.join().request(1);
    }
    return queue.take();
}
```
I think we should `upstream.request(1)` **regardless** of queue is empty or not due to timing issue on multi threads!

(Example of timing issue on multi threads)
> 1. thread 1: nextStream()
> 2. thread 1: queue is empty, so upstream.request(1)
> 3. thread 1: queue.take()..
> 4. upstream: onNext() -> queue.add(..);
> 5. thread 2: nextStream()
> 6. thread 2: queue is NOT empty so don't upstream.request(1)
> 7. thread 2: queue.take()..

thread 1, 2 are `queue.take()..` but `queue.add(..) `is called **only once** so one thread will be blocked forever 🤔